src/python/modules/ensembl/compara/plugins/TestFiles.py

```diff
@@ -0,0 +1,361 @@
+                subtree = reduce(operator.getitem, dirpath.split(os.sep), tree_dict)
+                subtree.update({name: {} for name in dirnames})
+                subtree['.'] = set(filenames)
+            tree_dict = tree_dict['']
+        return tree_dict
+
+    @staticmethod
+    def _prune_tree(root: Dict, paths: Union[str, List] = None, raise_err: bool = False) -> Dict:
+        """Returns a pruned directory tree that only includes the given paths.
+
+        Note:
+            If `paths` is empty, it returns `root`.
+
+        Args:
+            root: Directory tree.
+            paths: Relative directory path(s) to include.
+            raise_err: Raise exception flag.
+
+        Raises:
+            ValueError: if `raise_err` is True and any relative path is not part of the directory tree.
+
+        """
+        if not paths:
+            return root
+        # Make sure paths is a list
+        if isinstance(paths, str):
+            paths = [paths]
+        pruned_tree = {}  # type: Dict
+        for rel_path in paths:
+            # Normalize relative path to ensure reduce() finds it correctly
+            dirnames = os.path.normpath(rel_path).split(os.path.sep)
+            try:
+                tree_node = reduce(operator.getitem, dirnames, root)
+            except KeyError:
+                if raise_err:
+                    # Suppress exception context to display only the ValueError
+                    raise ValueError("Path '{}' not found in the directory tree".format(rel_path)) from None
+                continue
+            # Create/find the path in the pruned tree and attach the subtree
+            pruned_node = pruned_tree
+            for name in dirnames[:-1]:
+                pruned_node = pruned_node.setdefault(name, {})
+            pruned_node[dirnames[-1]] = tree_node
+        return pruned_tree
+
+    @staticmethod
+    def _eval_tree(root: Dict, test_func: Callable) -> List:
+        """Returns the files in the directory tree for which the test function returns False.
+
+        Args:
+            test_func: Test function to apply to each file. It has to match the following interface::
+                def test_func(file: str) -> bool:
+                    ...
+
+        """
+        nodes_left = [('', root)]
+        mismatches = []
+        while nodes_left:
+            rel_path, node = nodes_left.pop()
+            for dirname, dir_content in node.items():
+                if dirname != '.':
+                    # Append dirname to the list of directories left to evaluate
+                    nodes_left.append((os.path.join(rel_path, dirname), dir_content))
+                else:
+                    # Apply test_func() to each file
+                    for filename in dir_content:
+                        filepath = os.path.join(rel_path, filename)
+                        if not test_func(filepath):
+                            mismatches.append(filepath)
+        return mismatches
```

Comment:
What about adding a method to the DirectoryListing class that recursively traverses the tree (can be done via a stack like here, rather than recursive calls) and _yields_ all the paths ? Then `_eval_tree` becomes
```python
for path in root.traverse():
    if test_func(path):
        yield path
```

But it could perhaps be inlined, e.g.
```python
mismatches = [path for path in common.traverse() if cmp_file_size(path)]
```

Reply:
I had the code this way but then I realised only `_eval_tree` was using the `traverse` option, so I merged them both. I can see your proposal "cleans" the trick I use for `flatten`, so it may be worth considering this once we discuss the other bits of the `DirectoryListing` class.